tchannel-core/src/test/java/com/uber/tchannel/channels/ConnectionTest.java

```diff
@@ -135,15 +202,174 @@ public void testConnectionFailure() throws Exception {
+
+        int port = server.getListeningPort();
+
+        // create client
+        final TChannel client = new TChannel.Builder("client")
+            .setServerHost(host)
+            .build();
+        final SubChannel subClient = client.makeSubChannel("server");
+        client.listen();
+
+
+        final AtomicInteger counter = new AtomicInteger(0);
+        for (int i = 0; i < 10; i++) {
+            RawRequest req = new RawRequest.Builder("server", "echo")
+                .setHeader("title")
+                .setBody("hello")
+                .setTimeout(5)
+                .build();
+
+            ListenableFuture<RawResponse> future = subClient.send(
+                req,
+                host,
+                port
+            );
+
+            Futures.addCallback(future, new FutureCallback<RawResponse>() {
+                @Override
+                public void onSuccess(RawResponse response) {
+                    if (counter.incrementAndGet() < 10) {
+                        assertEquals(ErrorType.Timeout, response.getError().getErrorType());
+                    } else {
+                        assertEquals(ErrorType.NetworkError, response.getError().getErrorType());
+                        assertEquals("Connection was reset due to network error", response.getError().getMessage());
+                    }
+
+                    response.release();
+                }
+
+                @Override
+                public void onFailure(Throwable throwable) {
+                    // we should never reach here
+                    assertTrue(false);
+                    System.out.println(throwable.getMessage());
+                }
+            });
+        }
+
+        sleep(500);
+
+        // Connection should still work after reset
+        echoHandler.setDelayed(false);
+        RawRequest req = new RawRequest.Builder("server", "echo")
+            .setHeader("title")
+            .setBody("hello")
+            .setTimeout(20000)
+            .build();
+
+        ListenableFuture<RawResponse> future = subClient.send(
+            req,
+            host,
+            port
+        );
+
+        RawResponse res = future.get();
+        assertFalse(res.isError());
+        assertEquals("title", res.getHeader());
+        assertEquals("hello", res.getBody());
+
+        res.release();
+
+        client.shutdown();
+        server.shutdown();
     }
 
     protected  class EchoHandler implements RequestHandler {
         public boolean accessed = false;
+        private boolean delayed = false;
+
+        public EchoHandler() {}
+
+        public EchoHandler(boolean delayed) {
+            this.delayed = delayed;
+        }
+
+        public void setDelayed(boolean delayed) {
+            this.delayed = delayed;
+        }
 
         @Override
         public RawResponse handle(Request request) {
+
+            if (delayed) {
+                try {
+                    sleep(300);
+                } catch (Exception ex) {
```

Comment:
It is InterruptedException, not too much we can do. First, it is unlikely we get interrupted here. Second, even we get interrupted, the test will fail. So, no fasle positives.